src/Google.Api.Gax/Api/Gax/Bundler.cs

```diff
@@ -0,0 +1,329 @@
+        private readonly BundleComposer<TRequest, TResponse> _composer;
+
+        // Mutable state, all protected by "monitor".
+
+        // These are bundles which are not yet in flight.
+        private readonly Dictionary<TRequest, Bundle> waitingBundles;
+        // These are bundles which are in flight - we may want to wait for them all to complete,
+        // or cancel them all, etc.
+        private readonly LinkedList<Bundle> inFlightBundles;
+
+        /// <summary>
+        /// Used for testing, primarily: checks whether the current thread owns the monitor (lock)
+        /// which protects mutable state.
+        /// </summary>
+        internal bool InsideLock() => Monitor.IsEntered(_monitor);
+
+        public Bundler(
+            BundleSettings settings,
+            BundleComposer<TRequest, TResponse> composer,
+            ApiCallable<TRequest, TResponse> callable)
+        {
+            // TODO: Validation
+            _settings = GaxPreconditions.CheckNotNull(settings, nameof(settings)).Clone();
+            _settings.Validate(nameof(settings));
+            this._composer = GaxPreconditions.CheckNotNull(composer, nameof(composer));
+            this._callable = GaxPreconditions.CheckNotNull(callable, nameof(callable));
+            waitingBundles = new Dictionary<TRequest, Bundle>(composer.RequestComparer);
+            inFlightBundles = new LinkedList<Bundle>();
+        }
+
+        public Task<TResponse> Call(TRequest request, CancellationToken cancellationToken)
+        {
+            GaxPreconditions.CheckNotNull(request, nameof(request));
+            // We never need to worry about anything else modifying the request after this.
+            request = request.Clone();
+            bool sendImmediately;
+            bool newBundle;
+            Bundle bundle;
+            var clientRequest = new ClientCall(request);
+            var responseTask = clientRequest.ResponseCompletionSource.Task;
+            lock (_monitor)
+            {
+                // Needs to be in the lock as it modifies collection state
+                newBundle = !waitingBundles.TryGetValue(request, out bundle);
+                if (newBundle)
+                {
+                    bundle = new Bundle(clientRequest);
+                    waitingBundles[request] = bundle;
+                }
+                else
+                {
+                    _composer.AddRequestEntries(bundle.CombinedRequest, request);
+                    bundle.ClientCalls.Add(clientRequest);
+                }
+
+                // Needs to be in the lock as it reads bundle state
+                sendImmediately =
+                    (_settings.EntryCountThreshold != null && _composer.GetEntryCount(bundle.CombinedRequest) >= _settings.EntryCountThreshold) ||
+                    (_settings.RequestSizeThreshold != null && bundle.CombinedRequest.CalculateSize() >= _settings.RequestSizeThreshold);
```

Comment:
It'll work, but I'd rather not do the work for getting the entry count or the combined request size if it's going to be useless. Will add a comment to that effect.